File: MultiLanguageSandbox/src/test/deps/pascal/33.pas
{$ASSERTIONS ON}
(*
This problem introduces a new modulo operation, denoted by "⊕". Calculation of x ⊕ y:
- If x is not a multiple of y, the result is x mod y.
- Otherwise, if x is a multiple of y, divide x by y until it's no longer a multiple, resulting in x'.
  The result is then x' mod y.
For instance, 4⊕5=4, 20⊕5=4, and 100⊕5=4.
Given a prime number p and an integer n, compute the value of n! ⊕ p, where n! is the factorial of n.
    >>> newModuloFactorial(3, 7)
    6
*)

function newModuloFactorial(n, p: QWord): QWord;
var
  factorial, temp: QWord;
begin
  // Compute n!
  factorial := 1;
  for var i := 2 to n do
  begin
    factorial := factorial * i;
  end;

  // Apply the custom modulo operation ⊕
  temp := factorial;
  while temp mod p = 0 do
  begin
    temp := temp div p;
  end;
  
  Result := temp mod p;
end;
procedure TestNewModuloFactorial;
begin
  Assert(newModuloFactorial(3, 7) = 6);
  Assert(newModuloFactorial(10, 3) = 1);
  Assert(newModuloFactorial(11, 7) = 4);
  Assert(newModuloFactorial(45, 7) = 1);
  Assert(newModuloFactorial(14, 7) = 2);
  Assert(newModuloFactorial(1919, 10007) = 3152);
  Assert(newModuloFactorial(810, 10007) = 3679);
  Assert(newModuloFactorial(3, 7) = 6);
  Assert(newModuloFactorial(10, 3) = 1);
  Assert(newModuloFactorial(1, 2) = 1);
  Assert(newModuloFactorial(5, 11) = 10);
  Assert(newModuloFactorial(6, 13) = 5);
  Assert(newModuloFactorial(8, 17) = 13);
  Assert(newModuloFactorial(15, 19) = 16);
  Assert(newModuloFactorial(21, 23) = 1);
  Assert(newModuloFactorial(30, 29) = 28);
  Assert(newModuloFactorial(100, 101) = 100);
end;

begin
  TestNewModuloFactorial;
  WriteLn('All tests passed successfully.');
end.

File: MultiLanguageSandbox/src/test/deps/lua/33.lua

--[[
This problem introduces a new modulo operation, denoted by the symbol "⊕". When calculating x ⊕ y,
if x is not a multiple of y, the result is the remainder of x divided by y. Otherwise, divide x by y
repeatedly until x is no longer a multiple of y, denoting the final value as x'. Then the result is
the remainder of x' divided by y. For example, 4⊕5=4, 20⊕5=4, 100⊕5=4. Given a prime number p, there
will be several queries. For each query, an integer n is given, and you need to compute the value of
n! ⊕ p, where n! is the factorial of n, i.e., the product of all positive integers less than or
equal to n.

Examples:
    >>> newModuloFactorial(3, 7)
    6
]]--

function newModuloFactorial(n, p)
    -- Compute n!
    local factorial = 1
    for i = 1, n do
        factorial = factorial * i
    end
    
    -- Handle the case where factorial is 0 (n >= p and p is prime, but since p is prime and n < p, factorial mod p is non-zero)
    -- But for n >= p, factorial is divisible by p, so we need to remove all p factors.
    local x = factorial
    while x % p == 0 do
        x = x / p
    end
    
    return x % p
end
do
    assert(newModuloFactorial(3, 7) == 6)
    assert(newModuloFactorial(10, 3) == 1)
    assert(newModuloFactorial(11, 7) == 4)
    assert(newModuloFactorial(45, 7) == 1)
    assert(newModuloFactorial(14, 7) == 2)
    assert(newModuloFactorial(1919, 10007) == 3152)
    assert(newModuloFactorial(810, 10007) == 3679)
    assert(newModuloFactorial(3, 7) == 6)
    assert(newModuloFactorial(10, 3) == 1)
    assert(newModuloFactorial(1, 2) == 1)
    assert(newModuloFactorial(5, 11) == 10)
    assert(newModuloFactorial(6, 13) == 5)
    assert(newModuloFactorial(8, 17) == 13)
    assert(newModuloFactorial(15, 19) == 16)
    assert(newModuloFactorial(21, 23) == 1)
    assert(newModuloFactorial(30, 29) == 28)
    assert(newModuloFactorial(100, 101) == 100)
end